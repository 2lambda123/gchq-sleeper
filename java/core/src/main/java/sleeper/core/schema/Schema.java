/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    public Schema() {
        this.rowKeyFields = new ArrayList<>();
        this.sortKeyFields = new ArrayList<>();
        this.valueFields = new ArrayList<>();
    }

    private Schema(Builder builder) {
        rowKeyFields = new ArrayList<>(builder.validRowKeyFields());
        sortKeyFields = new ArrayList<>(builder.validSortKeyFields());
        valueFields = new ArrayList<>(builder.validValueFields());
        validateNoDuplicates(Stream.concat(rowKeyFields.stream(),
                Stream.concat(sortKeyFields.stream(), valueFields.stream())));
    }

    public static Builder builder() {
        return new Builder();
    }

    public void setRowKeyFields(List<Field> rowKeyFields) {
        validateRowKeys(rowKeyFields);
        this.rowKeyFields.clear();
        this.rowKeyFields.addAll(rowKeyFields);
    }

    public void setRowKeyFields(Field... rowKeyFields) {
        setRowKeyFields(Arrays.asList(rowKeyFields));
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return rowKeyFields.stream().map(Field::getType).map(t -> (PrimitiveType) t).collect(Collectors.toList());
    }

    public Field getSingleRowKeyField() {
        if (rowKeyFields.size() != 1) {
            throw new IllegalStateException("Cannot get single row key field, have " + rowKeyFields.size());
        }
        return rowKeyFields.get(0);
    }

    public void setSortKeyFields(List<Field> sortKeyFields) {
        validateSortKeys(sortKeyFields);
        this.sortKeyFields.clear();
        this.sortKeyFields.addAll(sortKeyFields);
    }

    public void setSortKeyFields(Field... sortKeyFields) {
        setSortKeyFields(Arrays.asList(sortKeyFields));
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return sortKeyFields.stream().map(Field::getType).map(t -> (PrimitiveType) t).collect(Collectors.toList());
    }

    public void setValueFields(List<Field> valueFields) {
        this.valueFields.clear();
        this.valueFields.addAll(valueFields);
    }

    public void setValueFields(Field... valueFields) {
        setValueFields(Arrays.asList(valueFields));
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<String> getRowKeyFieldNames() {
        return rowKeyFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getSortKeyFieldNames() {
        return sortKeyFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getValueFieldNames() {
        return valueFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getAllFieldNames() {
        List<String> allFieldNames = new ArrayList<>();
        allFieldNames.addAll(getRowKeyFieldNames());
        allFieldNames.addAll(getSortKeyFieldNames());
        allFieldNames.addAll(getValueFieldNames());
        return allFieldNames;
    }

    public List<Field> getAllFields() {
        List<Field> allFields = new ArrayList<>();
        allFields.addAll(rowKeyFields);
        allFields.addAll(sortKeyFields);
        allFields.addAll(valueFields);
        return allFields;
    }

    public Optional<Field> getField(String fieldName) {
        return getAllFields().stream()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }

        private List<Field> validRowKeyFields() {
            return validateRowKeys(rowKeyFields);
        }

        private List<Field> validSortKeyFields() {
            if (sortKeyFields == null) {
                return Collections.emptyList();
            } else {
                return validateSortKeys(sortKeyFields);
            }
        }

        private List<Field> validValueFields() {
            if (valueFields == null) {
                return Collections.emptyList();
            } else {
                return valueFields;
            }
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return fields;
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return fields;
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
