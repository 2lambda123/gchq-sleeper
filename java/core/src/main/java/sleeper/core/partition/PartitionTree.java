/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.schema.Schema;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * This represents a tree of {@link Partition}s. It can be used to find all
 * ancestors of a partition, i.e. all partitions that are either parents of the
 * partition, or grandparents, or great-grandparents.
 */
public class PartitionTree {
    private final Schema schema;
    private final Map<String, Partition> idToPartition;
    private final Partition rootPartition;

    public PartitionTree(Schema schema, List<Partition> partitions) {
        this.schema = schema;
        this.idToPartition = new HashMap<>();
        partitions.forEach(p -> this.idToPartition.put(p.getId(), p));
        List<Partition> rootPartitions = partitions.stream().filter(p -> null == p.getParentPartitionId()).collect(Collectors.toList());
        // There should be exactly one root partition.
        if (rootPartitions.size() != 1) {
            throw new IllegalArgumentException("There should be exactly one root partition, found " + rootPartitions.size());
        }
        this.rootPartition = rootPartitions.get(0);
    }

    public List<String> getChildIds(String partitionId) throws IllegalArgumentException {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return idToPartition.get(partitionId).getChildPartitionIds();
    }

    public String getParentId(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return idToPartition.get(partitionId).getParentPartitionId();
    }

    public List<String> getAllAncestorIds(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }

        List<String> ancestors = new ArrayList<>();
        String parent = idToPartition.get(partitionId).getParentPartitionId();
        while (null != parent) {
            ancestors.add(parent);
            parent = null == idToPartition.get(parent) ? null : idToPartition.get(parent).getParentPartitionId();
        }
        return ancestors;
    }

    public List<Partition> getAllAncestors(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }

        List<Partition> ancestors = new ArrayList<>();
        String parentId = idToPartition.get(partitionId).getParentPartitionId();
        while (null != parentId) {
            ancestors.add(idToPartition.get(parentId));
            parentId = idToPartition.get(parentId).getParentPartitionId();
        }
        return ancestors;
    }

    public Partition getPartition(String partitionId) {
        return idToPartition.get(partitionId);
    }

    public List<Partition> getAllPartitions() {
        return Collections.unmodifiableList(new ArrayList<>(idToPartition.values()));
    }

    public List<Partition> getLeafPartitions() {
        return Collections.unmodifiableList(idToPartition.values().stream()
                .filter(Partition::isLeafPartition)
                .collect(Collectors.toList()));
    }

    public Partition getLeafPartition(Key key) {
        // Sanity check key is of the correct length
        if (key.size() != schema.getRowKeyFields().size()) {
            throw new IllegalArgumentException("Key must match the row key fields from the schema (key was "
                    + key + ", schema has row key fields " + schema.getRowKeyFields() + ")");
        }
        // If root partition is a leaf partition then key must be in it
        if (rootPartition.isLeafPartition()) {
            return rootPartition;
        }

        return descend(rootPartition, key);
    }

    private Partition descend(Partition currentNode, Key key) {
        // Get child partitions
        List<String> childPartitionIds = currentNode.getChildPartitionIds();
        List<Partition> childPartitions = new ArrayList<>();
        childPartitionIds.forEach(c -> childPartitions.add(idToPartition.get(c)));

        // Which child is the key in?
        Partition child = null;
        for (Partition partition : childPartitions) {
            if (partition.isRowKeyInPartition(schema, key)) {
                child = partition;
                break;
            }
        }
        if (null == child) {
            throw new IllegalArgumentException("Found key that was not in any of the child partitions: key " + key
                    + ", child partitions " + childPartitions);
        }
        if (child.isLeafPartition()) {
            return child;
        }
        return descend(child, key);
    }

    public Partition getRootPartition() {
        return rootPartition;
    }

    public Partition getNearestCommonAncestor(Key a, Key b) {
        return getNearestCommonAncestor(getLeafPartition(a), getLeafPartition(b));
    }

    public Partition getNearestCommonAncestor(Partition a, Partition b) {
        if (a.getId().equals(b.getId())) {
            return a;
        }
        Set<String> ancestorsB = new HashSet<>(getAllAncestorIds(b.getId()));
        for (String ancestorA : getAllAncestorIds(a.getId())) {
            if (ancestorsB.contains(ancestorA)) {
                return getPartition(ancestorA);
            }
        }
        return getRootPartition();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PartitionTree that = (PartitionTree) o;
        return schema.equals(that.schema)
                && idToPartition.equals(that.idToPartition);
    }

    @Override
    public int hashCode() {
        return Objects.hash(schema, idToPartition);
    }
}
