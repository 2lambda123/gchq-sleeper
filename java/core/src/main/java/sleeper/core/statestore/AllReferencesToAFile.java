/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.core.statestore;

import java.time.Instant;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.TreeMap;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Reports on all the references for an individual physical file. A file may be referenced in a number of different
 * partitions, and may also have other external references which contribute to a combined reference count (eg. a
 * long-running query may count as a reference to the file).
 */
public class AllReferencesToAFile {

    private final String filename;
    private final Instant lastStateStoreUpdateTime;
    private final int totalReferenceCount;
    private final Map<String, FileReference> internalReferenceByPartitionId;

    private AllReferencesToAFile(Builder builder) {
        filename = Objects.requireNonNull(builder.filename, "filename must not be null");
        lastStateStoreUpdateTime = builder.lastStateStoreUpdateTime;
        totalReferenceCount = builder.totalReferenceCount;
        internalReferenceByPartitionId = Objects.requireNonNull(builder.internalReferenceByPartitionId, "internalReferenceByPartitionId must not be null");
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Creates a record for a file referenced in a single partition. This is used in state store implementations to
     * model
     * the file when only a reference was provided to add the file to the state store.
     *
     * @param  reference  the reference to the file
     * @param  updateTime the update time in the state store (should be set by the state store implementation)
     * @return            the file record
     */
    public static AllReferencesToAFile fileWithOneReference(FileReference reference, Instant updateTime) {
        return builder()
                .filename(reference.getFilename())
                .internalReferenceByPartitionId(Map.of(
                        reference.getPartitionId(),
                        reference.toBuilder().lastStateStoreUpdateTime(updateTime).build()))
                .totalReferenceCount(1)
                .lastStateStoreUpdateTime(updateTime)
                .build();
    }

    /**
     * Aggregates file references to create a record for each referenced file. This is used in state store
     * implementations to convert to the internal model when only references were provided to add files to the state
     * store. Every reference to each file must be included in the input, or the resulting model will be incorrect.
     *
     * @param  references references to files, including every reference to each file
     * @return            a stream of records for the referenced files
     */
    public static Stream<AllReferencesToAFile> newFilesWithReferences(Stream<FileReference> references) {
        Map<String, List<FileReference>> referencesByFilename = references
                .collect(Collectors.groupingBy(FileReference::getFilename, TreeMap::new, Collectors.toUnmodifiableList()));
        return referencesByFilename.entrySet().stream()
                .map(entry -> AllReferencesToAFile.builder()
                        .filename(entry.getKey())
                        .internalReferences(entry.getValue())
                        .totalReferenceCount(entry.getValue().size())
                        .build());
    }

    /**
     * Creates a copy of this model with the reference on one partition removed, and replaced with new references. This
     * is used in state store implementations to split a file reference into two to move the file down the tree of
     * partitions. The new references should cover all the records that were previously covered by the reference that's
     * being removed.
     *
     * @param  partitionId   the ID of the partition to remove the file from
     * @param  newReferences the references to add
     * @param  updateTime    the update time that this occurs (should be set by the state store implementation)
     * @return               a copy of the file record with this change applied
     */
    public AllReferencesToAFile splitReferenceFromPartition(
            String partitionId, Collection<FileReference> newReferences, Instant updateTime) {
        return toBuilder()
                .internalReferences(Stream.concat(
                        internalReferenceByPartitionId.values().stream()
                                .filter(reference -> !partitionId.equals(reference.getPartitionId())),
                        newReferences.stream().map(reference -> reference.toBuilder().lastStateStoreUpdateTime(updateTime).build())))
                .totalReferenceCount(totalReferenceCount - 1 + newReferences.size())
                .lastStateStoreUpdateTime(updateTime)
                .build();
    }

    /**
     * Creates a copy of this model with the reference on one partition removed. This is used in state store
     * implementations when adding the output of a compaction in a new file that contains all the records for a certain
     * partition. This means that the input files for the compaction must no longer be referenced in that partition.
     *
     * @param  partitionId the ID of the partition to remove the file from
     * @param  updateTime  the update time that this occurs (should be set by the state store implementation)
     * @return             a copy of the file record with this change applied
     */
    public AllReferencesToAFile removeReferenceForPartition(String partitionId, Instant updateTime) {
        return toBuilder()
                .internalReferences(internalReferenceByPartitionId.values().stream()
                        .filter(reference -> !partitionId.equals(reference.getPartitionId())))
                .totalReferenceCount(totalReferenceCount - 1)
                .lastStateStoreUpdateTime(updateTime)
                .build();
    }

    /**
     * Creates a copy of this model with the reference on one partition assigned to a job. This is used in state store
     * implementations when assigning a compaction job to its input files. Note that parts of a file are assigned to
     * jobs independently. Each partition that a file is in covers different records. Each reference on each
     * partition will be assigned to and processed by a different job.
     *
     * @param  jobId       the ID of the job to assign the file reference to
     * @param  partitionId the ID of the partition whose reference should be assigned to the job
     * @param  updateTime  the update time that this occurs (should be set by the state store implementation)
     * @return             a copy of the file record with this change applied
     */
    public AllReferencesToAFile withJobIdForPartition(String jobId, String partitionId, Instant updateTime) {
        return toBuilder()
                .internalReferences(internalReferenceByPartitionId.values().stream()
                        .map(reference -> {
                            if (partitionId.equals(reference.getPartitionId())) {
                                return reference.toBuilder().jobId(jobId).lastStateStoreUpdateTime(updateTime).build();
                            } else {
                                return reference;
                            }
                        }))
                .lastStateStoreUpdateTime(updateTime)
                .build();
    }

    /**
     * Creates a copy of this model with the update time set for the whole file and all its references. This is used in
     * state store implementations when adding a new file to the state store.
     *
     * @param  updateTime the update time that the file is added (should be set by the state store implementation)
     * @return            a copy of the file record with the update time set
     */
    public AllReferencesToAFile withCreatedUpdateTime(Instant updateTime) {
        return toBuilder()
                .internalReferences(internalReferenceByPartitionId.values().stream()
                        .map(reference -> reference.toBuilder().lastStateStoreUpdateTime(updateTime).build()))
                .lastStateStoreUpdateTime(updateTime)
                .build();
    }

    public String getFilename() {
        return filename;
    }

    public Instant getLastStateStoreUpdateTime() {
        return lastStateStoreUpdateTime;
    }

    public int getTotalReferenceCount() {
        return totalReferenceCount;
    }

    public int getExternalReferenceCount() {
        return totalReferenceCount - internalReferenceByPartitionId.size();
    }

    public Collection<FileReference> getInternalReferences() {
        return internalReferenceByPartitionId.values();
    }

    /**
     * Retrieves the reference for this file on a given partition.
     *
     * @param  partitionId the ID of the partition to find the reference in
     * @return             the reference to this file in the partition, if the file is referenced in that partition
     */
    public Optional<FileReference> getReferenceForPartitionId(String partitionId) {
        return Optional.ofNullable(internalReferenceByPartitionId.get(partitionId));
    }

    public Builder toBuilder() {
        return builder()
                .filename(filename)
                .internalReferenceByPartitionId(internalReferenceByPartitionId)
                .totalReferenceCount(totalReferenceCount)
                .lastStateStoreUpdateTime(lastStateStoreUpdateTime);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        AllReferencesToAFile that = (AllReferencesToAFile) o;
        return totalReferenceCount == that.totalReferenceCount && Objects.equals(filename, that.filename) && Objects.equals(lastStateStoreUpdateTime, that.lastStateStoreUpdateTime)
                && Objects.equals(internalReferenceByPartitionId, that.internalReferenceByPartitionId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(filename, lastStateStoreUpdateTime, totalReferenceCount, internalReferenceByPartitionId);
    }

    @Override
    public String toString() {
        return "AllReferencesToAFile{" +
                "filename='" + filename + '\'' +
                ", lastStateStoreUpdateTime=" + lastStateStoreUpdateTime +
                ", totalReferenceCount=" + totalReferenceCount +
                ", internalReferences=" + internalReferenceByPartitionId.values() +
                '}';
    }

    /**
     * Builder to create a file record.
     */
    public static final class Builder {
        private String filename;
        private Instant lastStateStoreUpdateTime;
        private int totalReferenceCount;
        private Map<String, FileReference> internalReferenceByPartitionId;

        private Builder() {
        }

        /**
         * Sets the filename.
         *
         * @param  filename the filename
         * @return          the builder
         */
        public Builder filename(String filename) {
            this.filename = filename;
            return this;
        }

        /**
         * Sets the time of the last state store update. Note that the update time is tracked separately for each
         * internal reference to the file as well. This should only be called by an implementation of the state store.
         *
         * @param  lastStateStoreUpdateTime the update time (should be set by the state store implementation)
         * @return                          the builder
         */
        public Builder lastStateStoreUpdateTime(Instant lastStateStoreUpdateTime) {
            this.lastStateStoreUpdateTime = lastStateStoreUpdateTime;
            return this;
        }

        /**
         * Sets the total count of references to the file. This should include internal references in partitions, and
         * any external references, eg. due to running queries reading the file. Note that for any new file this should
         * be equal to the number of internal references.
         *
         * @param  totalReferenceCount the total count of references
         * @return                     the builder
         */
        public Builder totalReferenceCount(int totalReferenceCount) {
            this.totalReferenceCount = totalReferenceCount;
            return this;
        }

        /**
         * Sets the internal references to the file on partitions.
         *
         * @param  references the references
         * @return            the builder
         */
        public Builder internalReferences(Stream<FileReference> references) {
            Map<String, FileReference> map = new TreeMap<>();
            references.forEach(reference -> map.put(reference.getPartitionId(), reference));
            return internalReferenceByPartitionId(Collections.unmodifiableMap(map));
        }

        /**
         * Sets the internal references to the file on partitions.
         *
         * @param  references the references
         * @return            the builder
         */
        public Builder internalReferences(Collection<FileReference> references) {
            return internalReferences(references.stream());
        }

        private Builder internalReferenceByPartitionId(Map<String, FileReference> internalReferenceByPartitionId) {
            this.internalReferenceByPartitionId = internalReferenceByPartitionId;
            return this;
        }

        public AllReferencesToAFile build() {
            return new AllReferencesToAFile(this);
        }
    }
}
