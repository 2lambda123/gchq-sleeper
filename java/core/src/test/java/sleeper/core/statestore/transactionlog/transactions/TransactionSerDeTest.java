/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.transactionlog.transactions;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionTree;
import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.StringType;
import sleeper.core.statestore.AllReferencesToAFile;
import sleeper.core.statestore.FileReference;
import sleeper.core.statestore.FileReferenceFactory;
import sleeper.core.statestore.transactionlog.FileReferenceTransactionGeneric;
import sleeper.core.statestore.transactionlog.StateStoreTransaction;
import sleeper.core.statestore.transactionlog.StateStoreTransactionGeneric;

import java.time.Instant;
import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toUnmodifiableList;
import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;
import static sleeper.core.statestore.AssignJobIdRequest.assignJobOnPartitionToFiles;
import static sleeper.core.statestore.SplitFileReference.referenceForChildPartition;
import static sleeper.core.statestore.SplitFileReferenceRequest.splitFileToChildPartitions;

public class TransactionSerDeTest {

    private static void whenSerDeThenMatchAndVerify(Schema schema, StateStoreTransactionGeneric<?> transaction) {
        // When
        TransactionSerDe serDe = new TransactionSerDe(schema);
        TransactionType type = TransactionType.getType(transaction);
        String json = serDe.toJsonPrettyPrint(transaction);

        // Then
        assertThat(serDe.toTransaction(type, json))
                .isEqualTo(transaction);
        Approvals.verify(json);
    }

    @Test
    void shouldSerDeAddFiles() {
        // Given
        Schema schema = schemaWithKey("key");
        PartitionTree partitions = new PartitionsBuilder(schema).singlePartition("root").buildTree();
        Instant updateTime = Instant.parse("2024-03-26T09:43:01Z");
        FileReferenceFactory fileFactory = FileReferenceFactory.fromUpdatedAt(partitions, updateTime);
        FileReferenceTransactionGeneric transaction = new AddFilesTransaction(
                AllReferencesToAFile.newFilesWithReferences(Stream.of(
                        fileFactory.rootFile("file1.parquet", 100),
                        fileFactory.rootFile("file2.parquet", 200)),
                        updateTime)
                        .collect(toUnmodifiableList()),
                updateTime);

        // When / Then
        whenSerDeThenMatchAndVerify(schema, transaction);
    }

    @Test
    void shouldSerDeAddSplitFile() {
        // Given
        Schema schema = schemaWithKey("key", new StringType());
        PartitionTree partitions = new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", "p")
                .buildTree();
        Instant updateTime = Instant.parse("2024-03-26T09:43:01Z");
        FileReferenceFactory fileFactory = FileReferenceFactory.fromUpdatedAt(partitions, updateTime);
        FileReference file = fileFactory.rootFile("file.parquet", 200);
        referenceForChildPartition(file, "L");
        FileReferenceTransactionGeneric transaction = new AddFilesTransaction(
                AllReferencesToAFile.newFilesWithReferences(Stream.of(
                        referenceForChildPartition(file, "L"),
                        referenceForChildPartition(file, "R")),
                        updateTime).collect(toUnmodifiableList()),
                updateTime);

        // When / Then
        whenSerDeThenMatchAndVerify(schema, transaction);
    }

    @Test
    void shouldSerDeAssignJobIds() {
        // Given
        FileReferenceTransactionGeneric transaction = new AssignJobIdsTransaction(List.of(
                assignJobOnPartitionToFiles("job1", "root",
                        List.of("file1.parquet", "file2.parquet")),
                assignJobOnPartitionToFiles("job2", "L",
                        List.of("file3.parquet", "file4.parquet"))),
                Instant.parse("2024-03-26T09:44:01Z"));

        // When / Then
        whenSerDeThenMatchAndVerify(schemaWithKey("key"), transaction);
    }

    @Test
    void shouldSerDeClearFiles() {
        // Given
        FileReferenceTransactionGeneric transaction = new ClearFilesTransaction();

        // When / Then
        whenSerDeThenMatchAndVerify(schemaWithKey("key"), transaction);
    }

    @Test
    void shouldSerDeDeleteFiles() {
        // Given
        FileReferenceTransactionGeneric transaction = new DeleteFilesTransaction(List.of("file1.parquet", "file2.parquet"));

        // When / Then
        whenSerDeThenMatchAndVerify(schemaWithKey("key"), transaction);
    }

    @Test
    void shouldSerDeInitialisePartitions() {
        // Given
        Schema schema = schemaWithKey("key", new StringType());
        StateStoreTransaction transaction = new InitialisePartitionsTransaction(new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", "p")
                .splitToNewChildren("L", "LL", "LR", "g")
                .splitToNewChildren("R", "RL", "RR", "u")
                .buildList());

        // When / Then
        whenSerDeThenMatchAndVerify(schema, transaction);
    }

    @Test
    void shouldSerDeReplaceFileReferences() throws Exception {
        // Given
        Schema schema = schemaWithKey("key");
        PartitionTree partitions = new PartitionsBuilder(schema).singlePartition("root").buildTree();
        Instant updateTime = Instant.parse("2023-03-26T10:05:01Z");
        FileReferenceFactory fileFactory = FileReferenceFactory.fromUpdatedAt(partitions, updateTime);
        FileReferenceTransactionGeneric transaction = new ReplaceFileReferencesTransaction(
                "job", "root", List.of("file1.parquet", "file2.parquet"),
                fileFactory.rootFile("file3.parquet", 100), updateTime);

        // When / Then
        whenSerDeThenMatchAndVerify(schema, transaction);
    }

    @Test
    void shouldSerDeSplitFileReferences() {
        // Given
        Schema schema = schemaWithKey("key", new StringType());
        PartitionTree partitions = new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", "p")
                .splitToNewChildren("L", "LL", "LR", "g")
                .splitToNewChildren("R", "RL", "RR", "u")
                .buildTree();
        Instant updateTime = Instant.parse("2023-03-26T10:05:01Z");
        FileReferenceFactory fileFactory = FileReferenceFactory.fromUpdatedAt(partitions, updateTime);
        FileReferenceTransactionGeneric transaction = new SplitFileReferencesTransaction(List.of(
                splitFileToChildPartitions(
                        fileFactory.rootFile("file1.parquet", 100), "L", "R"),
                splitFileToChildPartitions(
                        fileFactory.partitionFile("L", "file2.parquet", 200), "LL", "LR")),
                updateTime);

        // When / Then
        whenSerDeThenMatchAndVerify(schema, transaction);
    }

    @Test
    void shouldSerDeSplitPartition() {
        // Given
        Schema schema = schemaWithKey("key", new StringType());
        PartitionTree partitions = new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", "p")
                .splitToNewChildren("L", "LL", "LR", "g")
                .buildTree();
        StateStoreTransaction transaction = new SplitPartitionTransaction(
                partitions.getPartition("L"),
                List.of(partitions.getPartition("LL"), partitions.getPartition("LR")));

        // When / Then
        whenSerDeThenMatchAndVerify(schema, transaction);
    }
}
