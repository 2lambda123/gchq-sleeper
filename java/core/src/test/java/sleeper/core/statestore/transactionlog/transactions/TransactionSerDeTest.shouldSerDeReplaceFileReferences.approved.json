{
  "jobId": "job",
  "partitionId": "root",
  "inputFiles": [
    "file1.parquet",
    "file2.parquet"
  ],
  "newReference": {
    "filename": "file3.parquet",
    "partitionId": "root",
    "numberOfRecords": 100,
    "countApproximate": false,
    "onlyContainsDataForThisPartition": true
  },
  "updateTime": 1679825101000
}